lib/worldwide/deprecated_time_zone_mapper.rb

```diff
@@ -8,6 +10,103 @@ class << self
       def to_supported(zone)
         DEPRECATED_ZONES_MAP.fetch(zone&.to_sym, zone)
       end
+
+      # Map IANA timezones to Rails timezones
+      def to_rails(iana_timezone)
+        # If the timezone is already a supported Ruby timezone, return it
+        return iana_timezone if RAILS_TIMEZONES.key?(iana_timezone&.to_sym)
+
+        # IANA timezones has aliases. If the timezone is an alias, return the corresponding IANA timezone
+        timezone_alias = TIMEZONES_ALIASES.find do |timezone, aliases|
+          return timezone.to_s if aliases.include?(iana_timezone)
+        end
+
+        # If the alias is also a supported Rails timezone, return it
+        return timezone_alias if RAILS_TIMEZONES.key?(timezone_alias&.to_sym)
+
+        timezone = timezone_alias || iana_timezone
+
+        matching_timezones = find_matching_timezones(timezone)
+
+        return timezone if matching_timezones.empty?
+
+        iana_tz_info = TZInfo::Timezone.get(timezone)
+
+        # Find the closest Rails timezone based on geographical coordinates
+        closest_timezone = find_closest_timezone(iana_tz_info, matching_timezones)
```

Comment:
Unfortunately, geographic proximity does not correlate reliably with the timezone that's in effect.  For example, this code ends up mapping `America/North_Dakota/Beulah` (Central time, UTC-6, but with daylight saving observed at UTC-5) to its near-neighbour `America/Regina` (Central time, UTC-6, but does not observe daylight saving, so it's currently an hour different from Beulah as I type this in May 2024):

```ruby
shopify(dev)> Worldwide::DeprecatedTimeZoneMapper.to_rails("America/North_Dakota/Beulah")
=> "America/Regina"
```

I don't think there's a good substitute here for a manual review; fortunately, there are only a few hundred mappings to consider.  On the upside, we can get rid of the dependency on `geokit`, because it doesn't help us.  ;)

Reply:
Sorry for the silly question, but that's something I'd start to look how to solve. What should be considered in this case? `America/North_Dakota/Beulah` isn't a supported timezone and we should map to something else, but the closest timezone we could use is `America/Regina`. All the others are either to far away (e.g. `America/Bogota` and `America/Lima`) or also outside DST like `America/Regina`.

Reply:
> Sorry for the silly question, but that's something I'd start to look how to solve. What should be considered in this case? `America/North_Dakota/Beulah` isn't a supported timezone and we should map to something else, but the closest timezone we could use is `America/Regina`. All the others are either to far away (e.g. `America/Bogota` and `America/Lima`) or also outside DST like `America/Regina`.

That's a very good question (not silly at all!).  😄 

I think it helps to consider "why" we want a timezone.  Generally, it's so that we can convert a UTC date/time into a local date/time that the merchant or buyer will understand.  So, for example, if we expect a package to arrive in Toronto by 2024-05-29 at 19:00 UTC, we want to be able to display that in the UI as 2024-05-29 at 5pm.  To that end, for timezones that aren't supported by Rails, we want to find an alternative timezone that will give us the same formatted time result.  In this example, we know that we can use `America/New_York` instead of `America/Toronto`, because the two time zones have the same offsets and DST transition dates in 2024.

(David Olson made a design decision decades ago that tzdb would care about any timezones that presented a different time at any point after the start of the Unix "epoch" in 1970-01-01.  As a result, there are a lot of time zones in the IANA tzdb that only differ in what happened at some point shortly after 1970, and give equivalent answers for any "recent" date.  Continuing with the example above, `America/Toronto` and `America/New_York` have had the same offset and DST transition rules since 1986.) 

In the case of `America/North_Dakota/Beulah`, we need to identify an alternative timezone that
 - is supported by Rails,
 - has the same offsets and DST transition rules in 2024

The tzdb entry for Beulah is:
```
Zone America/North_Dakota/Beulah -6:47:07 - LMT 1883 Nov 18 19:00u
                        -7:00   US      M%sT    2010 Nov  7  2:00
                        -6:00   US      C%sT
```
This means that, from November 2010 to the present, Beulah uses offset `-6:00` with the `US C%sT` rules for DST transitions.

The zone `America/Chicago` has used offset `-6:00` and the `US C%sT` rules for DST transitions since 1967:

```
Zone America/Chicago    -5:50:36 -      LMT     1883 Nov 18 18:00u
                        -6:00   US      C%sT    1920
                        -6:00   Chicago C%sT    1936 Mar  1  2:00
                        -5:00   -       EST     1936 Nov 15  2:00
                        -6:00   Chicago C%sT    1942
                        -6:00   US      C%sT    1946
                        -6:00   Chicago C%sT    1967
                        -6:00   US      C%sT
```

Thus, for any date after 2010, we can safely use `America/Chicago` to localize UTC to the `America/North_Dakota/Beulah` time zone, and can list the former as the `.to_rails()` alias for the latter; it's one of the Rails-supported timezones.

```ruby
shopify(dev)> Worldwide::TimeZone.all.map(&:name).include?("America/Chicago")
=> true
```

To make this concrete:
 - 2024-05-29 19:00 UTC is 2024-05-29 at 4pm in Beulah
 - 2024-05-29 19:00 UTC is also 2024-05-29 at 4pm in Chicago
 - 2024-05-29 19:00 UTC is 2024-05-29 at 3pm in Regina

Therefore, `America/Chicago` is an acceptable alias for `America/North_Dakota/Beulah`, but `America/Regina` is not.